File: tf_quant_finance/math/interpolation/linear/linear_interpolation.cc
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <algorithm>
#include <cstddef>

#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/tensor_types.h"
#include "tensorflow/core/lib/core/errors.h"
#include "tensorflow/core/lib/core/threadpool.h"

namespace tensorflow {
namespace qf {

using CPUDevice = Eigen::ThreadPoolDevice;
using GPUDevice = Eigen::GpuDevice;

namespace functor {

template <typename Device, typename T>
struct LinearInterpolationFunctor {
  void operator()(OpKernelContext* context, typename TTypes<T>::ConstMatrix x,
                  typename TTypes<T>::ConstMatrix x_data,
                  typename TTypes<T>::ConstMatrix y_data,
                  typename TTypes<T>::ConstFlat left_slope,
                  typename TTypes<T>::ConstFlat right_slope,
                  typename TTypes<T>::Matrix output);
};

template <typename T>
struct LinearInterpolationFunctor<CPUDevice, T> {
  void operator()(OpKernelContext* context, typename TTypes<T>::ConstMatrix x,
                  typename TTypes<T>::ConstMatrix x_data,
                  typename TTypes<T>::ConstMatrix y_data,
                  typename TTypes<T>::ConstFlat left_slope,
                  typename TTypes<T>::ConstFlat right_slope,
                  typename TTypes<T>::Matrix output) {
    const size_t batch_size = x.dimension(0);
    const size_t x_size = x.dimension(1);
    const size_t x_data_size = x_data.dimension(1);

    // This lambda expression does not live beyond the scope of this function as
    // it is only passed to blocking calls.
    const auto compute = [&](int start, int end) {
      size_t index = start % x_size;
      size_t batch = start / x_size;

      for (size_t i = start; i < end; ++i) {
        const T* x_data_begin_ptr = x_data.data() + batch * x_data_size;
        const T* x_data_end_ptr = x_data.data() + (batch + 1) * x_data_size;

        const T curr_x = x(batch, index);

        // Represent lower and upper bound of x(index) in the sorted x_data.
        // These bounds can be used to distinguish between edge cases (e.g.
        // extrapolation).
        auto bounds =
            std::equal_range(x_data_begin_ptr, x_data_end_ptr, curr_x);

        if (bounds.first == x_data_begin_ptr) {
          // Left extrapolation case (including the edge case where the x[index]
          // is equal to the leftmost point in x_data).
          output(batch, index) =
              left_slope(batch) * (curr_x - x_data(batch, 0)) +
              y_data(batch, 0);
        } else if (bounds.second == x_data_end_ptr) {
          // Right extrapolation case.
          output(batch, index) =
              right_slope(batch) * (curr_x - x_data(batch, x_data_size - 1)) +
              y_data(batch, x_data_size - 1);
        } else {
          const size_t left_index = bounds.first - x_data_begin_ptr;
          const size_t right_index = bounds.second - x_data_begin_ptr;

          if (left_index == right_index) {
            // Interpolation step.
            const T x1 = x_data(batch, left_index - 1);
            const T x2 = x_data(batch, left_index);
            const T y1 = y_data(batch, left_index - 1);
            const T y2 = y_data(batch, left_index);
            output(batch, index) = (y2 - y1) / (x2 - x1) * (curr_x - x1) + y1;
          } else {
            // Element x(index) is present in the x_data.
            output(batch, index) = y_data(batch, left_index);
          }
        }

        index++;
        if (index == x_size) {
          index = 0;
          ++batch;
        }
      }
    };

    auto worker_threads = *(context->device()->tensorflow_cpu_worker_threads());

    const size_t num_elements = batch_size * x_size;
    const size_t num_threads = worker_threads.num_threads;

    // Each thread should process at least min_block_size elements.
    const size_t min_block_size = 32;
    const size_t block_size =
        std::max(num_elements / num_threads, min_block_size);

    worker_threads.workers->TransformRangeConcurrently(block_size, num_elements,
                                                       compute);
  }
};

}  // namespace functor

template <typename Device, typename T>
class LinearInterpolationOp : public OpKernel {
 public:
  explicit LinearInterpolationOp(OpKernelConstruction* context)
      : OpKernel(context) {}

  void Compute(OpKernelContext* context) override {
    const Tensor& x_tensor = context->input(0);
    const Tensor& x_data_tensor = context->input(1);
    const Tensor& y_data_tensor = context->input(2);

    OP_REQUIRES(context,
                x_tensor.dims() == 2 && x_data_tensor.dims() == 2 &&
                    y_data_tensor.dims() == 2,
                Status(error::INVALID_ARGUMENT,
                       "x, x_data and y_data must be of rank 2"));

    // The first dimension is batch size.
    auto x = x_tensor.matrix<T>();
    auto x_data = x_data_tensor.matrix<T>();
    auto y_data = y_data_tensor.matrix<T>();

    OP_REQUIRES(context, x_data.dimensions() == y_data.dimensions(),
                Status(error::INVALID_ARGUMENT,
                       "x_data and y_data must be of the same shape"));

    const size_t batch_size = x.dimension(0);

    OP_REQUIRES(context, x_data.dimension(0) == batch_size,
                Status(error::INVALID_ARGUMENT,
                       "x, x_data and y_data must have the same batch size"));

    const auto l_sizes = context->input(3).shape().dim_sizes();
    const auto r_sizes = context->input(4).shape().dim_sizes();

    OP_REQUIRES(context, l_sizes.size() == 1 && l_sizes[0] == batch_size,
                errors::InvalidArgument(
                    "The shape of left_slope needs to match batching size"));

    OP_REQUIRES(context, r_sizes.size() == 1 && r_sizes[0] == batch_size,
                errors::InvalidArgument(
                    "The shape of right_slope needs to match batching size"));

    const auto left_slope = context->input(3).flat<T>();
    const auto right_slope = context->input(4).flat<T>();

    // Create and allocate memory for the output tensor.
    Tensor* output_ptr = nullptr;

    OP_REQUIRES_OK(context, context->allocate_output(
                                0, context->input(0).shape(), &output_ptr));

    auto output = output_ptr->matrix<T>();

    functor::LinearInterpolationFunctor<Device, T>()(
        context, x, x_data, y_data, left_slope, right_slope, output);
  }
};

REGISTER_OP("LinearInterpolation")
    .Attr("T: {float, double} = DT_DOUBLE")
    .Input("x: T")
    .Input("x_data: T")
    .Input("y_data: T")
    .Input("left_slope: T")
    .Input("right_slope: T")
    .Output("y: T")
    .SetShapeFn([](shape_inference::InferenceContext* c) {
      c->set_output(0, c->input(0));
      return Status::OK();
    });

REGISTER_KERNEL_BUILDER(
    Name("LinearInterpolation").Device(DEVICE_CPU).TypeConstraint<double>("T"),
    LinearInterpolationOp<CPUDevice, double>);
REGISTER_KERNEL_BUILDER(
    Name("LinearInterpolation").Device(DEVICE_CPU).TypeConstraint<float>("T"),
    LinearInterpolationOp<CPUDevice, float>);

#if GOOGLE_CUDA

extern template struct functor::LinearInterpolationFunctor<GPUDevice, double>;
extern template struct functor::LinearInterpolationFunctor<GPUDevice, float>;

REGISTER_KERNEL_BUILDER(
    Name("LinearInterpolation").Device(DEVICE_GPU).TypeConstraint<double>("T"),
    LinearInterpolationOp<GPUDevice, double>);
REGISTER_KERNEL_BUILDER(
    Name("LinearInterpolation").Device(DEVICE_GPU).TypeConstraint<float>("T"),
    LinearInterpolationOp<GPUDevice, float>);

#endif  // GOOGLE_CUDA

}  // namespace qf
}  // namespace tensorflow

File: tf_quant_finance/math/interpolation/linear/linear_interpolation.py
# Lint as: python3
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Linear interpolation method."""

import tensorflow.compat.v2 as tf
from tf_quant_finance.math.interpolation import utils


def interpolate(x,
                x_data,
                y_data,
                left_slope=None,
                right_slope=None,
                validate_args=False,
                optimize_for_tpu=False,
                dtype=None,
                name=None):
  """Performs linear interpolation for supplied points.

  Given a set of knots whose x- and y- coordinates are in `x_data` and `y_data`,
  this function returns y-values for x-coordinates in `x` via piecewise
  linear interpolation.

  `x_data` must be non decreasing, but `y_data` don't need to be because we do
  not require the function approximated by these knots to be monotonic.

  #### Examples

  ```python
  x = [-10, -1, 1, 3, 6, 7, 8, 15, 18, 25, 30, 35]
  x_data = [-1, 2, 6, 8, 18, 30.0]
  y_data = [10, -1, -5, 7, 9, 20]

  result = linear_interpolation(x, x_data, y_data)
  # [ 10, 10, 2.66666667, -2, -5, 1, 7, 8.4, 9, 15.41666667, 20, 20]
  ```

  Args:
    x: x-coordinates for which we need to get interpolation. A N-D `Tensor` of
      real dtype. First N-1 dimensions represent batching dimensions.
    x_data: x coordinates. A N-D `Tensor` of real dtype. Should be sorted
      in non decreasing order. First N-1 dimensions represent batching
      dimensions.
    y_data: y coordinates. A N-D `Tensor` of real dtype. Should have the
      compatible shape as `x_data`. First N-1 dimensions represent batching
      dimensions.
    left_slope: The slope to use for extrapolation with x-coordinate smaller
      than the min `x_data`. It's a 0-D or N-D `Tensor`.
      Default value: `None`, which maps to `0.0` meaning constant extrapolation,
      i.e. extrapolated value will be the leftmost `y_data`.
    right_slope: The slope to use for extrapolation with x-coordinate greater
      than the max `x_data`. It's a 0-D or N-D `Tensor`.
      Default value: `None` which maps to `0.0` meaning constant extrapolation,
      i.e. extrapolated value will be the rightmost `y_data`.
    validate_args: Python `bool` that indicates whether the function performs
      the check if the shapes of `x_data` and `y_data` are equal and that the
      elements in `x_data` are non decreasing. If this value is set to `False`
      and the elements in `x_data` are not increasing, the result of linear
      interpolation may be wrong.
      Default value: `False`.
    optimize_for_tpu: A Python bool. If `True`, the algorithm uses one-hot
      encoding to lookup indices of `x_values` in `x_data`. This significantly
      improves performance of the algorithm on a TPU device but may slow down
      performance on the CPU.
      Default value: `False`.
    dtype: Optional tf.dtype for `x`, x_data`, `y_data`, `left_slope` and
      `right_slope`.
      Default value: `None` which means that the `dtype` inferred by TensorFlow
      is used.
    name: Python str. The name prefixed to the ops created by this function.
      Default value: `None` which maps to 'linear_interpolation'.

  Returns:
    A N-D `Tensor` of real dtype corresponding to the x-values in `x`.
  """
  name = name or "linear_interpolation"
  with tf.name_scope(name):
    x = tf.convert_to_tensor(x, dtype=dtype, name="x")
    dtype = dtype or x.dtype
    x_data = tf.convert_to_tensor(x_data, dtype=dtype, name="x_data")
    y_data = tf.convert_to_tensor(y_data, dtype=dtype, name="y_data")
    # Try broadcast batch_shapes
    x, x_data = utils.broadcast_common_batch_shape(x, x_data)
    x, y_data = utils.broadcast_common_batch_shape(x, y_data)
    x_data, y_data = utils.broadcast_common_batch_shape(x_data, y_data)

    batch_shape = x.shape.as_list()[:-1]
    if not batch_shape:
      x = tf.expand_dims(x, 0)
      x_data = tf.expand_dims(x_data, 0)
      y_data = tf.expand_dims(y_data, 0)

    if left_slope is None:
      left_slope = tf.constant(0.0, dtype=x.dtype, name="left_slope")
    else:
      left_slope = tf.convert_to_tensor(left_slope, dtype=dtype,
                                        name="left_slope")
    if right_slope is None:
      right_slope = tf.constant(0.0, dtype=x.dtype, name="right_slope")
    else:
      right_slope = tf.convert_to_tensor(right_slope, dtype=dtype,
                                         name="right_slope")
    control_deps = []
    if validate_args:
      # Check that `x_data` elements is non-decreasing
      diffs = x_data[..., 1:] - x_data[..., :-1]
      assertion = tf.compat.v1.debugging.assert_greater_equal(
          diffs,
          tf.zeros_like(diffs),
          message="x_data is not sorted in non-decreasing order.")
      control_deps.append(assertion)
      # Check that the shapes of `x_data` and `y_data` are equal
      control_deps.append(
          tf.compat.v1.assert_equal(tf.shape(x_data), tf.shape(y_data)))

    with tf.control_dependencies(control_deps):
      # Get upper bound indices for `x`.
      upper_indices = tf.searchsorted(x_data, x, side="left", out_type=tf.int32)
      x_data_size = x_data.shape.as_list()[-1]
      at_min = tf.equal(upper_indices, 0)
      at_max = tf.equal(upper_indices, x_data_size)
      # Create tensors in order to be used by `tf.where`.
      # `values_min` are extrapolated values for x-coordinates less than or
      # equal to `x_data[..., 0]`.
      # `values_max` are extrapolated values for x-coordinates greater than
      # `x_data[..., -1]`.

      values_min = tf.expand_dims(y_data[..., 0], -1) + left_slope * (
          x - tf.broadcast_to(
              tf.expand_dims(x_data[..., 0], -1), shape=tf.shape(x)))
      values_max = tf.expand_dims(y_data[..., -1], -1) + right_slope * (
          x - tf.broadcast_to(
              tf.expand_dims(x_data[..., -1], -1), shape=tf.shape(x)))

      # `tf.where` evaluates all branches, need to cap indices to ensure it
      # won't go out of bounds.
      lower_encoding = tf.math.maximum(upper_indices - 1, 0)
      upper_encoding = tf.math.minimum(upper_indices, x_data_size - 1)
      # Prepare indices for `tf.gather` or `tf.one_hot`
      # TODO(b/156720909): Extract get_slice logic into a common utilities
      # module for cubic and linear interpolation
      if optimize_for_tpu:
        lower_encoding = tf.one_hot(lower_encoding, x_data_size,
                                    dtype=dtype)
        upper_encoding = tf.one_hot(upper_encoding, x_data_size,
                                    dtype=dtype)
      def get_slice(x, encoding):
        if optimize_for_tpu:
          return tf.math.reduce_sum(tf.expand_dims(x, axis=-2) * encoding,
                                    axis=-1)
        else:
          return tf.gather(x, encoding, axis=-1, batch_dims=x.shape.rank - 1)
      x_data_lower = get_slice(x_data, lower_encoding)
      x_data_upper = get_slice(x_data, upper_encoding)
      y_data_lower = get_slice(y_data, lower_encoding)
      y_data_upper = get_slice(y_data, upper_encoding)

      # Nan in unselected branches could propagate through gradient calculation,
      # hence we need to clip the values to ensure no nan would occur. In this
      # case we need to ensure there is no division by zero.
      x_data_diff = x_data_upper - x_data_lower
      floor_x_diff = tf.where(at_min | at_max, x_data_diff + 1, x_data_diff)
      interpolated = y_data_lower + (x - x_data_lower) * (
          tf.math.divide_no_nan(y_data_upper - y_data_lower, floor_x_diff))

      interpolated = tf.where(at_min, values_min, interpolated)
      interpolated = tf.where(at_max, values_max, interpolated)
      if batch_shape:
        return interpolated
      else:
        return tf.squeeze(interpolated, 0)
